### PR TITLE
Issue #17727 Started refactoring uses of $db and $table 

### PR DESCRIPTION
**Pull request by soloistic**
### Description

Issue #17727 Started refactoring uses of $db and $table to use the Objects - DatabaseName and TableName

Changed uses related to Privileges

Created PR to know if changes are on the right track before refactoring more uses that spans multiple classes

Signed-off-by: Solomon Okunuga <soloistic1@gmail.com>

### CODE REVIEW

**Review by williamdes — APPROVED** [review]
The code look good. If this PR is not finished you can "Convert to draft" :)

**Comment by soloistic** [comment]
Okay, happy to move back to Draft... Just wondered about Mauricio's comment:

"you don't need to change all cases before opening a pull request, you can change a small piece of code, test if it's working and then open a pull request. That will make your PR easier to review and merge."

What's a reasonable amount for a small piece of code in this context? Should I go for say 15-20 files at a time? More? Less?

**Comment by williamdes** [comment]
> Okay, happy to move back to Draft... Just wondered about Mauricio's comment:
> 
> "you don't need to change all cases before opening a pull request, you can change a small piece of code, test if it's working and then open a pull request. That will make your PR easier to review and merge."
> 
> What's a reasonable amount for a small piece of code in this context? Should I go for say 15-20 files at a time? More? Less?

Oh no worries, it's just about if you want to change more things or not. For now I approved the code changes. It can be merged. But if you want to push more work convert it to a draft

**Comment by williamdes** [comment]
> What's a reasonable amount for a small piece of code in this context? Should I go for say 15-20 files at a time? More? Less?

By group of end functions like you did is good for me, let's wait some feedback from @MauricioFauth first

**Comment by soloistic** [comment]
> Oh no worries, it's just about if you want to change more things or not. For now I approved the code changes. It can be merged. But if you want to push more work convert it to a draft

Ahh right, makes sense now

File: libraries/classes/Server/Privileges.php [hunk 1]
```diff
@@ -1251,16 +1253,16 @@ public function getHtmlForAddUser($dbname)
...
     /**
-     * @param string $db    database name
-     * @param string $table table name
+     * @param DatabaseName $db    database name
+     * @param TableName $table table name
      *
      * @return array
      */
-    public function getAllPrivileges(string $db, string $table = ''): array
+    public function getAllPrivileges(DatabaseName $db, TableName $table = null): array
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1261]:
```suggestion
    public function getAllPrivileges(DatabaseName $db, ?TableName $table = null): array
```

File: libraries/classes/Server/Privileges.php [hunk 2]
```diff
@@ -1386,12 +1388,12 @@ private function getGlobalAndDatabasePrivileges(string $db): array
     }
 
     /**
-     * @param string $db    database name
-     * @param string $table table name
+     * @param DatabaseName $db    database name
+     * @param TableName $table table name
      *
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1393]: You can safely remove this redundant information.

```suggestion
```

File: libraries/classes/Server/Privileges.php [hunk 3]
```diff
@@ -1322,11 +1324,11 @@ private function getSpecificPrivilege(array $row): array
     }
 
     /**
-     * @param string $db database name
+     * @param DatabaseName $db database name
      *
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1328]: You can safely remove this redundant information.

```suggestion
```

File: libraries/classes/Server/Privileges.php [hunk 4]
```diff
@@ -1251,16 +1253,16 @@ public function getHtmlForAddUser($dbname)
     }
 
     /**
-     * @param string $db    database name
-     * @param string $table table name
+     * @param DatabaseName $db    database name
+     * @param TableName $table table name
      *
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1258]: You can safely remove this redundant information.

```suggestion
```

File: libraries/classes/Server/Privileges.php [hunk 5]
```diff
@@ -1376,7 +1378,7 @@ private function getGlobalAndDatabasePrivileges(string $db): array
             (
                 SELECT `User`, `Host`, ' . $listOfPrivileges . ' `Db`, \'d\' AS `Type`
                 FROM `mysql`.`db`
-                WHERE \'' . $this->dbi->escapeString($db) . '\' LIKE `Db` AND NOT (' . $listOfComparedPrivileges . ')
+                WHERE \'' . $this->dbi->escapeString($db->getName()) . '\' LIKE `Db` AND NOT (' . $listOfComparedPrivileges . ')
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1381]: The length of the line exceeds 120 characters.
```suggestion
                WHERE \'' . $this->dbi->escapeString($db->getName()) . '\' LIKE `Db` AND NOT ('
                . $listOfComparedPrivileges . ')
```

File: libraries/classes/Server/Privileges.php [hunk 6]
```diff
@@ -1414,16 +1416,16 @@ private function getTablePrivileges(string $db, string $table): array
     }
 
     /**
-     * @param string $db database name
+     * @param DatabaseName $db database name
      *
```
Comment (MauricioFauth) [inline libraries/classes/Server/Privileges.php:1420]: You can safely remove this redundant information.

```suggestion
```

File: libraries/classes/Controllers/Table/PrivilegesController.php [hunk 7]
```diff
@@ -78,11 +80,11 @@ public function __invoke(ServerRequest $request): void
 
         $scriptName = Util::getScriptNameForOption($GLOBALS['cfg']['DefaultTabTable'], 'table');
 
-        $db = $GLOBALS['db'];
-        $table = $GLOBALS['table'];
+        $db = DatabaseName::fromValue($GLOBALS['db']);
+        $table = TableName::fromValue($GLOBALS['table']);
```
Comment (MauricioFauth) [inline libraries/classes/Controllers/Table/PrivilegesController.php:84]: I forgot to mention in the issue, but when an invalid name is provided, the `fromValue` method will throw an exception. This should be handled. 
However I added in 6792b17e72f28ec61a0a9b32055c751d34336e5e a new method called `tryFromValue` that returns `null` for invalid names.
`$GLOBALS['db']` and `$GLOBALS['table']` will hold an empty string for invalid names. So if `tryFromValue` returns `null`, you can assign an empty string (`$db !== null ? $db->getName() : ''`) as it's the current behavior.

**Review by MauricioFauth — CHANGES_REQUESTED** [review]
Thank you for your contribution.

File: libraries/classes/Server/Privileges.php (hunk 1 above)
Comment (williamdes) [reply libraries/classes/Server/Privileges.php:1261]: i did think about that but maybe it's better we can not pass this param as null so we use the simpler syntax each time ?

Comment (MauricioFauth) [reply libraries/classes/Server/Privileges.php]: > i did think about that but maybe it's better we can not pass this param as null so we use the simpler syntax each time ?

It does not prevent passing `null` (See https://3v4l.org/8Jsqv). But it fail the coding standard check. https://github.com/phpmyadmin/phpmyadmin/actions/runs/3078113001/jobs/4973518771#step:8:36

**Comment by MauricioFauth** [comment]
Merged, thanks for your contribution!

**Comment by soloistic** [comment]
Thank you!
And apologies I couldn't return to finish it up. I got a bit swamped with other responsibilities.

**Comment by MauricioFauth** [comment]
> Thank you! And apologies I couldn't return to finish it up. I got a bit swamped with other responsibilities.

You don't have to apologize. Thank you for you contribution.
I added the error handling and some tests in 135cd76d3b3d4a2ce805434632b286c915ad6dc9.

**Pull request merged**